_episodes/02-hardware-performance.md

```diff
@@ -21,177 +21,179 @@ best option is to explore parallel computing. Conceptually this is simple: split
-more surprisingly you will discover that in LAMMPS (**FIXME, should this be more generic??**), the MPI parallelisation is so strongly
-integrated that it almost always offers better performance than the OpenMP based thread-level
-parallelism unless your system has a fair amount of load-imbalance.
- 
+In addition to MPI, some applications also support thread level parallelism (through, for
+example OpenMP directives) which can
+offer additional parallelisation within a subdomain. The basic working principle in OpenMP
+is based the "Fork-Join" model, as shown below.
+
 <p align="center"><img src="../fig/02/fork_join.png" width="50%"/></p>
 
-Another major acceleration method which is offered by LAMMPS is through the graphical processing
-units (GPUs). GPUs work together with the CPUs. While a CPU is specialized to perform complex tasks
-like flow control, emulating virtual machines, security, etc., a GPU is very efficient in doing
-simple repetitive low level tasks. This functional difference between a GPU and CPU could be
-attributed to the massively parallel architecture that a GPU possesses. A modern CPU has only a
-few cores which are well optimized for performing sequential serial jobs. On the contrary, a GPU
-has thousands of cores that are highly efficient to do simple repetitive jobs in parallel. See
-below on how a CPU and GPU works together.
+In the 'Fork-Join' model, there exists a master thread which "fork"s into multiple threads.
+Each of these forked-threads executes a part of the whole job and when all the threads are
+done with their assigned jobs, these threads join together again. Typically, the number of
+threads is equal to the number of available cores, but this can be influenced by the
+application or the user at run-time.
 
-<p align="center"><img src="../fig/02/CPUplusGPU.png" width="50%"/></p> 
+Using the maximum possible threads on a node may not always provide
+the best performance. It depends on many factors
+and, in some cases,the MPI parallelization strategy is so strongly
+integrated that it almost always offers better performance than the OpenMP based thread-level
+parallelism.
+
+Another commonly found parallelization strategy is to use *GPU*s (with the more general
+term being an *accelerator*). GPUs work together with the CPUs. A CPU is specialized to perform
+complex tasks (like making decisions), while a GPU is very efficient in doing
+simple, repetitive, low level tasks. This functional difference between a GPU and CPU could be
+attributed to the *massively parallel* architecture that a GPU possesses. A modern CPU has
+relatively few cores which are well optimized for performing sequential serial jobs. On the
+contrary, a GPU
+has thousands of cores that are highly efficient at doing simple repetitive jobs in parallel. See
+below on how a CPU and GPU works together.
 
+<p align="center"><img src="../fig/02/CPUplusGPU.png" width="50%"/></p>
```

Comment:
This image needs work. Drop "ALU" and replace with "Core". Change DRAM for RAM. Swap arrow colours so that the bulk of work is sent to GPU. There's also a spelling mistake (repetitive).

It introduces the word cache which we haven't explained...

Reply:
@prithwish-ichec